NRSE documentation (#488)
PowerGridModel/power-grid-model: docs/user_manual/calculations.md

#### State estimation

State estimation is a statistical calculation method that determines the most probable state of the grid, based on

> network data and measurements. Here, measurements may be power flow or voltage values with some kind of uncertainty, which were 

Reviewer: Now that I re-read:
```suggestion
network data and measurements. Here, measurements can be power flow or voltage values with certain kind of uncertainty, which were 
```

- Power flow through branches
- Deviation between measurement values and estimated state

In order to perform a state estimation, the system should be observable. If the system is not observable, the calculation will 
raise a sparse matrix error, since the matrix reprensentation of the system is too sparse to solve - in particular when it is

> singular. Simply said, observability means that the number of measurements should be larger than or equal to the number of
> unknowns. For each node, there are two unknowns, `u` and `u_angle`, so the following equations should be met:

Reviewer: ```suggestion
singular. In short, meeting the requirement of observability indicates that the system is either an overdetermined system (when the number of measurements is larger than the number of
unknowns) or a balanced system (when the number of measurements is equal to the number of unknowns). For each node, there are two unknowns, `u` and `u_angle`, so the following conditions should be met:
```

- two times the number of branches with a power sensor

```{note}
Having enough measurements does not necessarily mean that the system is observable. The location of the measurements is

> also of importance. Also, there should be at least one voltage measurement.

Reviewer: ```suggestion
also of importance. Additionally, there should be at least one voltage measurement.
```

```

```{warning}
The [iterative linear](#iterative-linear-state-estimation) and [Newton-Raphson](#newton-raphson-state-estimation) state estimation algorithms will assume angles to be zero by default (see the details about voltage sensors).

> This produces more correct outputs when the system is observable, but will prevent the calculation from raising an exception, even if it is unobservable, therefore giving faulty results.

Reviewer: ```suggestion
In observable systems this helps better outputting correct results. On the other hand with unobservable systems, exceptions raised from calculations due to faulty results will be prevented.
```

| -------------------------------------------------- | -------- | -------- | -------- | ----------------------------------------------------------------------------------------------------------- |
| [Newton-Raphson](#newton-raphson-power-flow)       | &#10004; |          | &#10004; | {py:class}`CalculationMethod.newton_raphson <power_grid_model.enum.CalculationMethod.newton_raphson>`       |
| [Iterative current](#iterative-current-power-flow) |          |          | &#10004; | {py:class}`CalculationMethod.iterative_current <power_grid_model.enum.CalculationMethod.iterative_current>` |
| [Linear](#linear-power-flow)                       |          | &#10004; |          | {py:class}`CalculationMethod.linear <power_grid_model.enum.CalculationMethod.linear>`                       |
| [Linear current](#linear-current-power-flow)       |          | &#10004; |          | {py:class}`CalculationMethod.linear_current <power_grid_model.enum.CalculationMethod.linear_current>`       |

```{note}
By default, the [Newton-Raphson](#newton-raphson-power-flow) method is used.
```

```{note}

> When all the load/generation types are of constant impedance, power-grid-model uses the [Linear](#linear-power-flow) method regardless of the input provided by the user. 
> This is because this method will then be both accurate and the fastest.

Reviewer: ```suggestion
When all the load/generation types are of constant impedance, the [Linear](#linear-power-flow) method will be the fastest without loss of accuracy. Therefore power-grid-model will use this method regardless of the input provided by the user in this case. 
```

- Using LU decomposition, solve $J(i) \Delta x(i)  =  \Delta y(i)$ for $\Delta x(i)$
- Compute $x(i+1)$ from $\Delta x(i) =  x(i+1) - x(i)$

#### Iterative current power flow

Algorithm call: {py:class}`CalculationMethod.iterative_current <power_grid_model.enum.CalculationMethod.iterative_current>`

This algorithm is a Jacobi-like method for powerflow analysis.

> It has linear convergence as opposed to quadratic convergence in the [Newton-Raphson](#newton-raphson-power-flow) method. This means that the number of iterations will be greater.

Reviewer: ```suggestion
It has a linear convergence rate as opposed to the quadratic convergence rate in the [Newton-Raphson](#newton-raphson-power-flow) method. This means that more iterations is needed to achieve similar convergence.
```

> [Newton-Raphson](#newton-raphson-power-flow) will also be more robust in achieving convergence in case of greater meshed configurations. However, the iterative current algorithm will be faster most of the time.

Reviewer: ```suggestion
Additionally, [Newton-Raphson](#newton-raphson-power-flow) will be more robust converging in case of greater meshed configurations. Nevertheless, the iterative current algorithm will be faster most of the time.
```

Reviewer: This is a problematic sentence: `meshed configuration` is a term but why `greater`? Unless this is a domain terminology, I suggest we come up with clearer wording; the `iterative current` faster in what sense? In terms of the runtime performance per iteration or in achieving similar convergence?

The $Y_{bus}$ matrix here does not change across iterations which means it only needs to be factorized once to solve the linear equations in all iterations. 
The $Y_{bus}$ matrix also remains unchanged in certain batch calculations like timeseries calculations.

#### Linear power flow

Algorithm call: {py:class}`CalculationMethod.linear <power_grid_model.enum.CalculationMethod.linear>`

This is an approximation method where we assume that all loads and generations are of constant impedance type regardless of their actual {py:class}`LoadGenType <power_grid_model.enum.LoadGenType>`.
By doing so, we obtain huge performance benefits as the computation required is equivalent to a single iteration of the iterative methods.
It will be more accurate when most of the load/generation types are of constant impedance or the actual node voltages are close to 1 p.u.
When all the load/generation types are of constant impedance, power-grid-model uses Linear method regardless of the input provided by the user.

> This is because this method will then be both accurate and the fastest. 

Reviewer: See comment above.

The algorithm is as follows:

1. Assume injected currents by loads $I_N=0$ since we model loads/generation as impedance.
   Admittance of each load/generation is calculated from rated power as $y=-\overline{S}$
2. Build $Y{bus}$. Add the admittances of loads/generation to the diagonal elements.
3. Solve linear equation: $YU_N = I_N$ for $U_N$

#### Linear current power flow

Algorithm call: {py:class}`CalculationMethod.linear_current <power_grid_model.enum.CalculationMethod.linear_current>`

**This algorithm is essentially a single iteration of [iterative current power flow](#iterative-current-power-flow).**

This approximation method will give better results when most of the load/generation types resemble constant current.
Similar to [iterative current](#iterative-current-power-flow), batch calculations like timeseries will also be faster.

> The reason is the same: the $Y_{bus}$ matrix does not change across batches and the same factorization would be used.

Reviewer: ```suggestion
The reason is the same: the $Y_{bus}$ matrix does not change across batches and as a result the same factorization could be used.
```

Where $S_k$ and $\sigma_{P,k}$ and $\sigma_{Q,k}$ are the measured value and the standard deviation of the individual appliances.

#### Iterative linear state estimation

Algorithm call: {py:class}`CalculationMethod.iterative_linear <power_grid_model.enum.CalculationMethod.iterative_linear>`

> Linear WLS requires all measurements to be linear. This is only possible if all measurements are phasor unit measurements,

Reviewer: ```suggestion
Linear WLS requires all measurements to be linear. This is only possible when all measurements are phasor unit measurements,
```

> which is not realistic in a distribution grid. Therefore, traditional measurements are linearized before the algorithm is performed:

Reviewer: ```suggestion
which is not realistic in distribution grids. Therefore, traditional measurements are linearized prior to running the algorithm:
```

  - Normalize the voltage phasor angle by setting the angle of the slack bus to zero:
  - If the maximum deviation between $\underline{U}^{(k)}$ and $\underline{U}^{(k-1)}$ is smaller than the error tolerance $\epsilon$,
  stop the iteration. Otherwise, continue until the maximum number of iterations is reached. 
  
In the iteration process, the phase angle of voltages at each bus is updated using the last iteration;
the system error of the phase shift converges to zero. Because the matrix is pre-built and
pre-factorized, the computation cost of each iteration is much smaller than for the [Newton-Raphson](#newton-raphson-state-estimation)
method, where the Jacobian matrix needs to be constructed and factorized each time.

```{warning}
The algorithm will assume angles to be zero by default (see the details about voltage sensors). This produces more correct outputs when the system is observable, but will prevent the calculation from raising an exception, even if it is unobservable, therefore giving faulty results.
```

#### Newton-Raphson state estimation

```warning

> This feature is still experimental and is at the time of writing not yet publicly available.

Reviewer: ```suggestion
 At the time of writing this feature is still experimental and is not yet publicly available.
```

```

Algorithm call: {py:class}`CalculationMethod.newton_raphson <power_grid_model.enum.CalculationMethod.newton_raphson>`

> The Newton-Raphson approach solves state estimation by considering it as a system of real, non-linear equations.

Reviewer: ```suggestion
The Newton-Raphson state estimation considers the problem as a system of real, non-linear equations.
```

> It then iteratively solves the first order Taylor expansion of that system. It does not make any linearization assumptions.

Reviewer: ```suggestion
It iteratively solves the first order Taylor expansion of the system. It does not make any assumptions on linearization.
```

> It may therefore be more accurate and is likely to achieve convergence in less iterations than the [iterative linear](#iterative-linear-state-estimation) approach.

Reviewer: ```suggestion
It could be more accurate and converge in less iterations than the [iterative linear](#iterative-linear-state-estimation) approach.
```

Reviewer: The reason that it could be more accurate is unclear from context, hence removing `Therfore`.

> However, the Jacobian matrix needs to be calculated every iteration and cannot be prefactorized, which, on average, causes it to be slower.

Reviewer: ```suggestion
However, the Jacobian matrix needs to be calculated every iteration and cannot be prefactorized, which, on average, makes it slower.
```

> The Newton-Raphson methode considers all measurements to be independent real measurements.

Reviewer: ```suggestion
The Newton-Raphson method considers all measurements to be independent real measurements.
```

I.e., $\sigma_P$, $\sigma_Q$ and $\sigma_U$ are all independent values.
The rationale behind to calculation is similar to that of the [Newton-Raphson for power flow](#newton-raphson-power-flow).

> Because of that, the iteration process differs slightly from the one for [iterative linear state estimation](#iterative-linear-state-estimation), as shown below.

Reviewer: ```suggestion
Consequently, the iteration process differs slightly from that of [iterative linear state estimation](#iterative-linear-state-estimation), as shown below.
```

  - For Bus $i$, if there is a voltage measurement, $U_i^{(0)} = U_{\text{mea},i}$ and $\theta_i^{(0)} = \theta_i$,
  where $U_{\text{mea},i}$ is the measured voltage magnitude.
- In iteration $k$, follow the steps below:
  - Construct and (LU-)factorize the Jacobian matrix for $\boldsymbol{x}^{(k)}$, using the network parameters.
  - Compute the temporary new voltage magnitude and angle result, $\widetilde{\boldsymbol{U}}^{(k)}$ and $\widetilde{\boldsymbol{\theta}}^{(k)}$,
  using the prefactorized matrix. See also [Matrix-prefactorization](./performance-guide.md#matrix-prefactorization)
  - Normalize the result voltage phasor angle by setting angle of slack bus to zero:
  $\underline{U}_i^{(k)} = \widetilde{\underline{U}}_i^{(k)} \times |\widetilde{\underline{U}}_s^{(k)}| / \widetilde{\underline{U}}_s^{(k)}$.
  - If the maximum deviation between $\underline{\boldsymbol{U}}^{(k)}$ and $\underline{\boldsymbol{U}}^{(k-1)}$ is smaller than the tolerance $\epsilon$,
  stop the iteration, otherwise continue until the maximum number of iterations is reached.
  Note: we're using the phasor here: $\underline{\boldsymbol{U}} = \boldsymbol{U}e^{j\boldsymbol{\theta}}$.

> As for the [iterative linear](#iterative-linear-state-estimation) approach, in the iteration process, the phase angle of voltages at each bus is updated using the last iteration.

Reviewer: ```suggestion
As for the [iterative linear](#iterative-linear-state-estimation) approach, during iterations, phase angles of voltage at each bus are updated using ones from the previous iteration.
```

The system error of the phase shift converges to zero.

```{warning}
The algorithm will assume angles to be zero by default (see the details about voltage sensors). This produces more correct outputs when the system is observable, but will prevent the calculation from raising an exception, even if it is unobservable, therefore giving faulty results.

> ```

Reviewer: See above comment.
